pymc3/tests/test_distributions_random.py

```diff
@@ -498,43 +425,177 @@ class TestMoyal(BaseTestCases.BaseTestCase):
     params = {"mu": 0.0, "sigma": 1.0}
 
 
-@pytest.mark.skip(reason="This test is covered by Aesara")
-class TestCategorical(BaseTestCases.BaseTestCase):
-    distribution = pm.Categorical
-    params = {"p": np.ones(BaseTestCases.BaseTestCase.shape)}
+class TestCorrectParametrizationMappingPymcToScipy(SeededTest):
+    @staticmethod
+    def get_inputs_from_apply_node_outputs(outputs):
+        parents = outputs.get_parents()
+        if not parents:
+            raise Exception("Parent Apply node missing for output")
+        # I am assuming there will always only be 1 Apply parent node in this context
+        return parents[0].inputs
 
-    def get_random_variable(
-        self, shape, with_vector_params=False, **kwargs
-    ):  # don't transform categories
-        return super().get_random_variable(shape, with_vector_params=False, **kwargs)
+    def _compare_pymc_sampling_with_aesara_one(
+        self,
+        pymc_params,
+        expected_aesara_params,
+        pymc_dist,
+        size=15,
+        decimal=6,
+        sampling_aesara_params=None,
+    ):
+        pymc_params.append(("size", size))
+        pymc_dist_output = pymc_dist.dist(
+            rng=aesara.shared(self.get_random_state()), **dict(pymc_params)
+        )
+        self._pymc_params_match_aesara_rv_ones(pymc_dist_output, expected_aesara_params, decimal)
+        if sampling_aesara_params is None:
+            sampling_aesara_params = expected_aesara_params
+        self._pymc_sample_matches_aeasara_rv_one(
+            pymc_dist_output, pymc_dist, sampling_aesara_params, size, decimal
+        )
 
-    def test_probability_vector_shape(self):
-        """Check that if a 2d array of probabilities are passed to categorical correct shape is returned"""
-        p = np.ones((10, 5))
-        assert pm.Categorical.dist(p=p).random().shape == (10,)
-        assert pm.Categorical.dist(p=p).random(size=4).shape == (4, 10)
-        p = np.ones((3, 7, 5))
-        assert pm.Categorical.dist(p=p).random().shape == (3, 7)
-        assert pm.Categorical.dist(p=p).random(size=4).shape == (4, 3, 7)
+    def _pymc_params_match_aesara_rv_ones(
+        self, pymc_dist_output, expected_aesara_params, decimal=6
+    ):
+        aesera_dist_inputs = self.get_inputs_from_apply_node_outputs(pymc_dist_output)[3:]
+        assert len(expected_aesara_params) == len(aesera_dist_inputs)
+        for (expected_name, expected_value), actual_variable in zip(
+            expected_aesara_params, aesera_dist_inputs
+        ):
+            assert_almost_equal(expected_value, actual_variable.eval(), decimal=decimal)
+
+    def _pymc_sample_matches_aeasara_rv_one(
+        self, pymc_dist_output, pymc_dist, expected_aesara_params, size, decimal
+    ):
+        sample = pymc_dist.rv_op(
```

Comment:
I am not sure of what I am about to say at all, but here it goes:

We might want to go a bit more low level here and compare with the samples from the underlying "reference" number generating function (say `rng.exponential` or `scipy.stats.distribution.rvs()`) directly. Otherwise we are just comparing that the call from the `random op` and `pymc_dist.eval()` are the same, which is well covered by the `_pymc_params_match_aesara...`. 

Silly example: if we changed our exponential to return the outcome of `rng.exponential(mu+1)` instead of `rng.exponential(mu)` this test would not capture the accidental change. However, if we pass the reference function explicitly and call it with the "expected parameters" and the same seed, we would find that the outcomes don't match anymore.

As an example of how to define the reference random distribution see here: https://github.com/pymc-devs/pymc3/blob/b8edd5671cecf719d7059e3d728182db66c209bd/pymc3/tests/test_distributions_random.py#L819-L827

Basically duplicating the code that generates the random numbers...

It doesn't provide much more guarantees, but could make this test more useful for the remaining distributions that have not been refactored yet.

Does this make sense at all to you?

Reply:
I agree with you in principle, but I am not sure what's the best way to achieve that outcome in practice. Below my 2 cents.

The way I see it Aesara is responsible for generating the correct distribution given the rv class name. Indirectly Aesara is also responsible for testing that this is the case. PyMC relies on Aesara doing a good job at this, in the same way that Aesara relies on Scipy/Numpy doing a good job at implementation correctly the various distributions' methods (ie I assume these are not tested for correctness in Aesara, maybe I am wrong though).

I agree with you that adding a lower level test here will increase our confidence that everything works as expected, but I feel that probably these tests should be added to Aesara rather than here. Particularly so, given that the PyMC dev team owns that library, so this should be easy to achieve.

Also, if we decide that we want to have this extra level of safety in PyMC we should probably not rely on knowing Aesara implementation. So, rather than seeding Aesara and our random generators and checking for exact matches in a small sample, we should probably still rely on Kolmogorov Smirnov and Chi-Square tests. ie I wouldn't make the PyMC tests rely on Aesara current implementation (`np/scipy`), which may change in the future, making the tests fail, but still lead to the same result (ie right distribution).

To make sure I get your point fully:
> Silly example: if we changed our exponential to return the outcome of rng.exponential(mu+1) instead of rng.exponential(mu) this test would not capture the accidental change

Wouldn't this be caught by the parameterization test?

> It doesn't provide much more guarantees, but could make this test more useful for the remaining distributions that have not been refactored yet.

I guess that you would like each distribution to have a test similar to that, right?

Reply:
> We might want to go a bit more low level here and compare with the samples from the underlying "reference" number generating function (say rng.distribution or scipy.stats.distribution.rvs()) directly. 

Btw if you think a straightforward approach like the one you mentioned would work fine, I am happy to go for it

Reply:
Thanks for your input. I agree that for the aesara ops it's a complete overkill. I was hinting at the future random ops that will be implemented mostly on the PyMC3 side. If this test framework would also cover well these future cases it could be worthwhile, even if for the few ops that come from aesara it is indeed an overkill.

But maybe let's leave this as is and tackle future changes if they prove necessary later.

Reply:
>I was hinting at the future random ops that will be implemented mostly on the PyMC3 side

Oh, I see, I didn't know that. What's the best way to stay up to date with this kind of knowledge?

Is this something similar to what you had in mind (haven't ran it yet, so there maybe bugs/typos)? or would you still test it through `pm.Exponential.dist(...)`
```python
from aesara.tensor.random import basic as a_rv_ops

class TestRandomVariableDistribution(SeededTest):

    def _compare_pymc_sampling_with_aesara(self, rv_op_dist, rv_op_params, ref_dist, ref_params, decimal=6, size=15):
        pymc_dist_output = rv_op_dist(*rv_op_params.values(), size=size, rng=aesara.shared(self.get_random_state()))
        sample = ref_dist(self.get_random_state(), **ref_params, size=15)
        assert_array_almost_equal(pymc_dist_output.eval(), sample, decimal=decimal)

    def test_exponential_distribution(self):
        rv_op_params, ref_params = {"lam": 10.}, {"scale": 10.}
        rv_op_dist, ref_dist = a_rv_ops.exponential, np.random.RandomState.exponential
        self._compare_pymc_sampling_with_aesara(rv_op_params, rv_op_params, ref_dist, ref_params)
```

Reply:
Also, at this point I feel like the `_compare_pymc_sampling_with_aesara_one` method resulted from me misinterpreting the following `asserting that the numbers generated by our distribution and the scipy one were the same when given the same random seeds. That feels like a stronger test.`. Would it make sense to remove it?

Reply:
> Is this something similar to what you had in mind (haven't ran it yet, so there maybe bugs/typos)? or would you still test it through pm.Exponential.dist(...)

Yes exactly, does it make sense? I would still test through the pymc.dist as users won't be using random ops directly, unless you think this way is significantly easier or there are advantages that I am ignoring.

Reply:
> I guess that for now I can only add these tests for rv ops defined in aesara.

If you want to consider the future ops, you can look at the gumbel random op that was implemented recently in aesara as that calls scipy.stats directly. Most future random ops in pymc will follow that format (ie 1 to 1 correspondence to a scipy distribution rvs call). 

For something a tiny bit more complicated you can check the discrete weibull that I refactored in the draft PR. That makes use of random routines in numpy or scipy but is not as simple as calling a pre defined distribution.

If your tests could cover those 3 types of random ops (calling numpy directly, calling scipy directly, or building upon these more indirectly) it would probably fit most needs.

Reply:
> What's the best way to stay up to date with this kind of knowledge?

It's not always easy, even for core developers as things are moving quite fast right now.

In general some info can be gathered from the meeting minutes: https://github.com/pymc-devs/pymc3/wiki/Project-Meeting-Minutes

Or the wiki more broadly.

Hopefully things will become more predictable soon! It's always fine to ask of course :)

Sorry for the confusion and again thanks for your help and patience!

Reply:
Cool, thanks for the comments and pointing me to the info about where to find the meeting minutes.

I refactored the tests. Now it should be easy to add any extra tests for each distribution. I also added a test comparing the `.dist` outcome with the one coming from the reference distribution. For now it supports scipy and numpy distributions, it shouldn't be too hard to use a different callable to test a custom distribution (like the discrete Weibull you mentioned).

Let me know what you think about the new format. Also, sorry for asking you to re-review the code for this PR basically from scratch